### PR TITLE
New Plugin: Encounter

### PR DESCRIPTION
**Pull request by Daztek**
Adds a bunch of encounter functions:

GetNumberOfCreaturesInEncounterList()
GetEncounterCreatureByIndex()
SetEncounterCreatureByIndex()
GetFactionId()
SetFactionId()
GetPlayerTriggeredOnly()
SetPlayerTriggeredOnly()
GetResetTime()
SetResetTime()

### CODE REVIEW

**Comment by Daztek** [comment]
Should be good for now.

Added:

SetEncounterCreatureByIndex()
GetFactionId()
SetFactionId()
GetPlayerTriggeredOnly()
SetPlayerTriggeredOnly()

**Comment by Christian-Vassallo** [comment]
I love this plugin.

One of the main problem of nwn is that you can’t manage encounter by script and make things different than normal system, example random spawn.
You need tricks that can be complex.

Would it be possible to add/remove creature in the encounter using resref, according to gs?

Example:
AddCreatureToEncounter(object oEncounter, string resrefCreature, int Index, int GS)
And remove also
?

**Comment by Christian-Vassallo** [comment]
This was my post many months ago

https://github.com/nwnxee/unified/issues/152

**Comment by Daztek** [comment]
Adding/Removing is too much of a pain for me to add, it's editing existing entries only, I'm afraid. 

But you can technically remove entries by setting its challenge rating to something like 500

**Comment by Christian-Vassallo** [comment]
SetEncounterCreatureByIndex()

What is the aim of this function?

**Comment by Daztek** [comment]
You can edit existing entries of the encounter creature list with it.

Say you have an encounter with 4 creatures, you could change the second entry to a different creature like so:

```
void main()
{
    object oEncounter = GetObjectByTag("Encounter");

    struct NWNX_Encounter_CreatureListEntry creatureEntry;

    creatureEntry.resref = "nw_beardire";
    creatureEntry.challengeRating = 9.0f;
    creatureEntry.unique = FALSE;

    NWNX_Encounter_SetEncounterCreatureByIndex(oEncounter, 1, creatureEntry);
}
```

File: Plugins/Encounter/Encounter.cpp [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+
+    if (auto *pEncounter = encounter(args))
+    {
+        const auto index = Services::Events::ExtractArgument<int32_t>(args);
+        const auto resref = Services::Events::ExtractArgument<std::string>(args);
+        auto cr = Services::Events::ExtractArgument<float>(args);
+        auto unique = Services::Events::ExtractArgument<int32_t>(args);
+
+        if (cr < 0.0) cr = 0.0;
+        if (unique < 0) unique = 0;
```
[the remark this answers is not in the document]
Comment (Daztek) [reply Plugins/Encounter/Encounter.cpp]: Fixed

**Comment by Christian-Vassallo** [comment]
It sounds good. It’s a good way to do “add” a new creature. Just put the max number of creature you would like with dummy resref and then change them as you wish

**Pull request merged**